swupd/delta.go

```diff
@@ -31,149 +43,123 @@ func CreateDeltas(manifest, statedir string, from, to uint32) ([]*File, error) {
-			file.Info, _ = os.Stat(fullname)
-			if file.Info.Size() > 200 {
-				files = append(files, file)
-			}
-		}
-
+	// Check that the delta actually applies correctly.
+	testPath := filepath.Join(deltaDir, "."+deltaName+".testnewfile")
+	if err := helpers.RunCommandSilent("bspatch", oldPath, testPath, deltaPath); err != nil {
+		return errors.Wrapf(err, "Failed to apply delta %s", deltaPath)
 	}
-
-	return files, nil
-}
-
-func createDelta(c config, file *File, statedir string, fromVersion uint32, fromHash Hashval) error {
-	// File without peers cannot have deltas
-	if file.DeltaPeer == nil {
-		return nil
+	testHash, err := Hashcalc(testPath)
+	if err != nil {
+		return errors.Wrap(err, "Failed to calculate hash for test file created applying delta")
 	}
-
-	// We only support deltas between two regular files for now
-	// TODO: Support directory -> file in the future, this is a complex case
-	if file.Type != TypeFile || file.DeltaPeer.Type != TypeFile {
-		return nil
+	if testHash != delta.To.Hash {
+		return errors.Wrapf(err, "Delta mismatch: %s -> %s via delta: %s", oldPath, newPath, deltaPath)
 	}
+	_ = os.Remove(testPath)
 
-	newfile := filepath.Join(c.imageBase, fmt.Sprintf("%d", file.Version), "full", file.Name)
-	original := filepath.Join(c.imageBase, fmt.Sprintf("%d", fromVersion), "full", file.Name)
-
-	deltadir := filepath.Join(c.outputDir, fmt.Sprintf("%d", file.Version), "delta")
-	fromToString := fmt.Sprintf("%d-%d-%s-%s", fromVersion, file.Version, fromHash, file.Hash)
-
-	// Files to create and validate deltas with
-	deltafile := filepath.Join(deltadir, fromToString)
-	testnewfile := filepath.Join(deltadir, "."+fromToString+".testnewfile")
-
-	// Shell out to bsdiff for now...false = don't print to screen
-	if err := helpers.RunCommandSilent("bsdiff", original, newfile, deltafile); err != nil {
-		_ = os.Remove(deltafile) // Might have returned FULLDL
-		return errors.Wrapf(err, "Failed to create delta for %s -> %s", original, newfile)
-	}
+	delta.path = deltaPath
+	return nil
+}
 
-	// Check that the delta actually applies correctly
-	if err := helpers.RunCommandSilent("bspatch", original, testnewfile, deltafile); err != nil {
-		return errors.Wrapf(err, "Failed to apply delta %s", deltafile)
-	}
+// Delta represents a delta file between two other files. If Error is present, it
+// indicates that the delta couldn't be created.
+type Delta struct {
+	From  *File
+	To    *File
+	Error error
 
-	deltahash, err := Hashcalc(testnewfile)
-	if err != nil {
-		return errors.Wrap(err, "Failed to calculate hash for new file")
-	}
+	path string
+}
 
-	if !HashEquals(deltahash, file.Hash) {
-		return errors.Wrapf(err, "Delta mismatch: %s -> %s via delta: %s", original, newfile, deltafile)
+func findDeltas(c *config, oldManifest, newManifest *Manifest) ([]Delta, error) {
+	// TODO: Pre-allocate with the maximum size possible (wasteful) or trust the reallocs?
```

Comment:
I'd say trust the reallocs. This is already very memory intensive so we probably don't want to assume worst case here and pre-allocate to the max size.